### PR TITLE
FINERACT-1269: (fix) Validate the json we try to deserialize is valid.

### PR DESCRIPTION
**Pull request by Zeyad2003**
## Description

- Wrap the fromJson() method with try-catch block and throw proper exception.
- Add `publishHookEventHandlesInvalidJson()` method to test the invalid json is handled successfully.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by adamsaghy** [comment]
@Zeyad2003 Please follow the proper PR and commit message requirements!
You can take an example from your other PR: https://github.com/apache/fineract/pull/3804

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Also please dont forget to add proper tests to cover your fix

**Comment by Zeyad2003** [comment]
> @Zeyad2003 Please follow the proper PR and commit message requirements! You can take an example from your other PR: #3804

Noted.

**Comment by Zeyad2003** [comment]
> Also please dont forget to add proper tests to cover your fix

Do you mean to add some unit-tests to `isValidJson()` method or something else?

**Comment by adamsaghy** [comment]
> > Also please dont forget to add proper tests to cover your fix
> 
> Do you mean to add some unit-tests to `isValidJson()` method or something else?

Yes,

You should test your fix by providing valid and invalid json and check whether it was handled and proper exception was raised.

For that you can write unit test and/or integration test where you are testing one of the API and check what happens when invalid json body was provided.

Regards
Adam

**Comment by adamsaghy** [comment]
@Zeyad2003 Please follow the proper PR and commit message requirements!
Please squash your commits!
The rule is 1 PR, 1 commit!

You can take an example from your other PR: #3804

**Comment by Zeyad2003** [comment]
> @Zeyad2003 Please follow the proper PR and commit message requirements! Please squash your commits! The rule is 1 PR, 1 commit!
> 
> You can take an example from your other PR: #3804

Done.
Sorry, I thought we should put the testing and the other implementation in separate commits.

File: fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java [hunk 1]
```diff
@@ -272,7 +272,12 @@ private void publishHookEvent(final String entityName, final String actionName,
                 Type type = new TypeToken<Map<String, Object>>() {
 
                 }.getType();
-                Map<String, Object> myMap = gson.fromJson(command.json(), type);
+
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java:297]: This is not correct... if the provided json is invalid, you should raise a PlatformApiDataValidationException with some message of the nature of the exception, like: "Invalid JSON was provided" or something like this. Feel free to check how the above exception was raised in Fineract and with what parameters.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Kindly check my review!

File: fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java (hunk 1 above)
Comment (Zeyad2003) [reply fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java:297]: Do you mean replacing the if condition to be something like that:

```java
if (!isValidJson(command.json(), type)) {
    throw new PlatformApiDataValidationException("error.msg.invalid.json", "The provided JSON is invalid.", new ArrayList<>());
}
```

If there's something else please let me know.

File: fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java [hunk 2]
```diff
@@ -272,6 +272,11 @@ private void publishHookEvent(final String entityName, final String actionName,
                 Type type = new TypeToken<Map<String, Object>>() {
 
                 }.getType();
+
+                if (!isValidJson(command.json(), type)) {
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java]: This method is not required. I would just simply wrap the `Map<String, Object> myMap = gson.fromJson(command.json(), type);` around with try catch and throw the `PlatformApiDataValidationException` exception.

File: fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java [hunk 3]
```diff
@@ -293,7 +298,12 @@ private void publishHookEvent(final String entityName, final String actionName,
                 } else if (result instanceof ErrorInfo ex) {
                     reqmap.put("status", "Exception");
 
+                    if (!isValidJson(ex.getMessage(), type)) {
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java]: This is wrong... you should do the following:
wrap around the `gson.fromJson(ex.getMessage(), type)`
```
Map<String, Object> errorMap;
try {
       errorMap = gson.fromJson(ex.getMessage(), type);
} catch (Exception e) {
        errorMap.put("errorMessage", ex.getMessage());
  }
errorMap.put("errorCode", ex.getErrorCode());
errorMap.put("statusCode", ex.getStatusCode());
   ```

File: fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java (hunk 2 above)
Comment (adamsaghy) [reply fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java]: This way we can avoid double json parsing

File: fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java (hunk 3 above)
Comment (Zeyad2003) [reply fineract-core/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java]: Okay, I think this makes more sense.
But for testing, does the existing tests are enough or do we need to add some other tests? (given that `publishHookEvent()` is private)

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**